File: offer/63.go
/*
 * @Author: lmio
 * @Date: 2023-03-29 10:11:49
 * @LastEditTime: 2023-04-07 14:32:50
 * @FilePath: /leetcode/offer/63.go
 * @Description:剑指 Offer 63. 股票的最大利润
 */
package offer

import (
	"leetcode/utils"
	"math"
)

func MaxProfit(prices []int) int {
	cost := math.MaxInt
	profit := math.MinInt
	for _, price := range prices {
		cost = utils.Min(cost, price)
		profit = utils.Max(profit, price - cost)
	}
	return profit
}

File: code/63.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-07-15 20:37:22
LastEditors: lmio 2091319361@qq.com
Description: 
'''
from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        n, m = len(obstacleGrid), len(obstacleGrid[0])
        dp =  [[0 for _ in range(m)] for _ in range(n)]
        dp[0][0] = obstacleGrid[0][0]^1
        for j in range(1, m):
            dp[0][j] = dp[0][j-1] & (obstacleGrid[0][j]^1)
        for i in range(1, n):
            dp[i][0] = dp[i-1][0] & (obstacleGrid[i][0]^1)
        for i in range(1, n):
            for j in range(1, m):
                if obstacleGrid[i][j] == 1:
                    dp[i][j] = 0
                    continue
                dp[i][j] = dp[i][j-1] + dp[i-1][j]
        return dp[n-1][m-1]